File: python/app/core/progress.py
import sys
sys.dont_write_bytecode = True

import threading
import time

_lock  = threading.Lock()
_state = {
    "active": False, "total": 0, "done": 0, "current": "",
    "percent": 0, "phase": "idle", "errors": 0,
    "eta_sec": -1, "elapsed": 0, "file_types": {},
}

_phase_start_time:    float = 0.0
_done_at_phase_start: int   = 0


def set_progress(done=None, total=None, current=None, phase=None, errors=None):
    global _phase_start_time, _done_at_phase_start
    with _lock:
        if phase is not None and phase != _state["phase"]:
            _phase_start_time    = time.time()
            _done_at_phase_start = done if done is not None else _state["done"]

        if done    is not None: _state["done"]    = done
        if total   is not None: _state["total"]   = total
        if current is not None: _state["current"] = current
        if phase   is not None: _state["phase"]   = phase
        if errors  is not None: _state["errors"]  = errors

        if _state["total"] > 0:
            _state["percent"] = round(_state["done"] / _state["total"] * 100, 1)

        _state["active"]  = _state["phase"] != "idle"
        elapsed           = time.time() - _phase_start_time
        _state["elapsed"] = round(elapsed, 1)
        items_done        = _state["done"] - _done_at_phase_start
        items_remaining   = _state["total"] - _state["done"]

        if elapsed > 2 and items_done > 0 and items_remaining > 0:
            _state["eta_sec"] = round(items_remaining / (items_done / elapsed))
        else:
            _state["eta_sec"] = -1


def set_file_type_totals(counts: dict):
    with _lock:
        _state["file_types"] = {ext: {"done": 0, "total": count} for ext, count in counts.items()}


def increment_file_type(ext: str):
    with _lock:
        ext = ext.lower().lstrip(".")
        if ext in _state["file_types"]:
            _state["file_types"][ext]["done"] += 1
        else:
            _state["file_types"][ext] = {"done": 1, "total": 1}


def increment_errors():
    with _lock:
        _state["errors"] += 1


def get_progress() -> dict:
    with _lock:
        snap = dict(_state)
        snap["file_types"] = {k: dict(v) for k, v in _state["file_types"].items()}
        return snap


def reset():
    global _phase_start_time, _done_at_phase_start
    with _lock:
        _phase_start_time = 0.0
        _done_at_phase_start = 0
        _state.update({
            "active": False, "total": 0, "done": 0, "current": "",
            "percent": 0, "phase": "idle", "errors": 0,
            "eta_sec": -1, "elapsed": 0, "file_types": {},
        })

File: UI/src-tauri/src/core/progress.rs
//! Thread-safe progress state shared between the sync/search worker thread
//! and the Tauri command that polls it every 500 ms.

use once_cell::sync::Lazy;
use serde::Serialize;
use std::{
    collections::HashMap,
    sync::Mutex,
    time::Instant,
};

// ── Public snapshot type (sent to Angular) ────────────────────────────────

#[derive(Debug, Clone, Serialize)]
pub struct ProgressSnapshot {
    pub active:      bool,
    pub phase:       String,
    pub done:        usize,
    pub total:       usize,
    pub current:     String,
    pub percent:     f32,
    pub eta_sec:     i64,
    pub errors:      usize,
    pub elapsed:     f32,
    pub file_types:  HashMap<String, FileTypeCount>,
}

#[derive(Debug, Clone, Serialize)]
pub struct FileTypeCount {
    pub done:  usize,
    pub total: usize,
}

// ── Internal mutable state ────────────────────────────────────────────────

struct ProgressState {
    phase:              String,
    done:               usize,
    total:              usize,
    current:            String,
    errors:             usize,
    file_types:         HashMap<String, FileTypeCount>,
    phase_start:        Option<Instant>,
    done_at_phase_start: usize,
}

impl Default for ProgressState {
    fn default() -> Self {
        Self {
            phase:               "idle".into(),
            done:                0,
            total:               0,
            current:             String::new(),
            errors:              0,
            file_types:          HashMap::new(),
            phase_start:         None,
            done_at_phase_start: 0,
        }
    }
}

static STATE: Lazy<Mutex<ProgressState>> =
    Lazy::new(|| Mutex::new(ProgressState::default()));

// ── Public API ────────────────────────────────────────────────────────────

pub fn set_progress(
    phase:   Option<&str>,
    done:    Option<usize>,
    total:   Option<usize>,
    current: Option<&str>,
    errors:  Option<usize>,
) {
    let mut s = STATE.lock().unwrap();

    if let Some(p) = phase {
        if p != s.phase {
            s.phase_start        = Some(Instant::now());
            s.done_at_phase_start = done.unwrap_or(s.done);
        }
        s.phase = p.to_string();
    }
    if let Some(v) = done    { s.done    = v; }
    if let Some(v) = total   { s.total   = v; }
    if let Some(v) = current { s.current = v.to_string(); }
    if let Some(v) = errors  { s.errors  = v; }
}

pub fn increment_errors() {
    STATE.lock().unwrap().errors += 1;
}

pub fn set_file_type_totals(counts: HashMap<String, usize>) {
    let mut s = STATE.lock().unwrap();
    s.file_types = counts
        .into_iter()
        .map(|(k, total)| (k, FileTypeCount { done: 0, total }))
        .collect();
}

pub fn increment_file_type(ext: &str) {
    let ext = ext.trim_start_matches('.').to_lowercase();
    let mut s = STATE.lock().unwrap();
    let entry = s.file_types.entry(ext).or_insert(FileTypeCount { done: 0, total: 1 });
    entry.done += 1;
}

pub fn get_progress() -> ProgressSnapshot {
    let s = STATE.lock().unwrap();

    let percent = if s.total > 0 {
        (s.done as f32 / s.total as f32 * 100.0 * 10.0).round() / 10.0
    } else {
        0.0
    };

    let (eta_sec, elapsed) = match s.phase_start {
        None => (-1_i64, 0.0_f32),
        Some(start) => {
            let elapsed  = start.elapsed().as_secs_f32();
            let items_done      = s.done.saturating_sub(s.done_at_phase_start);
            let items_remaining = s.total.saturating_sub(s.done);

            let eta = if elapsed > 2.0 && items_done > 0 && items_remaining > 0 {
                (items_remaining as f32 / (items_done as f32 / elapsed)).round() as i64
            } else {
                -1
            };
            (eta, (elapsed * 10.0).round() / 10.0)
        }
    };

    ProgressSnapshot {
        active:     s.phase != "idle",
        phase:      s.phase.clone(),
        done:       s.done,
        total:      s.total,
        current:    s.current.clone(),
        percent,
        eta_sec,
        errors:     s.errors,
        elapsed,
        file_types: s.file_types.clone(),
    }
}

pub fn reset() {
    let mut s = STATE.lock().unwrap();
    *s = ProgressState::default();
}
